Unified cache maintenance services for OP-TEE core

etienne-lms opened the pull request:
Provides cache maintenance service to OP-TEE core. In current OP-TEE core, unified cache maintenance services are only provided to TAs through a SVC syscall. This PR aims at providing the feature to both OP-TEE core and OP-TEE TA.

This PR is extracted from https://github.com/OP-TEE/optee_os/pull/1322 and supersedes it.

etienne-lms commented:
This change is required by the SDP feature, to allow pseudo TAs to flush/clean/invalidate SDP memory when a SDP pTA is invoked to access (read and/or write) a SDP buffer.

Actually, these changes are useful only to the platforms relying on a outer cache.

jforissier commented on core/arch/arm/tee/cache.c:
```diff
@@ -49,13 +49,15 @@ TEE_Result cache_operation(enum utee_cache_operation op, void *va, size_t len)
 
 	switch (op) {
 	case TEE_CACHEFLUSH:
+#ifdef CFG_PL310
```
I think this is correct, but I am concerned with the terminology. I mean, `cache_maintenance_l2()` does nothing when `CFG_PL310` is not set, yet the platform is very likely to have a L2 cache. I find it confusing. I think we should introduce "outer" in the function name. How about `cache_maintenance_outer_l2()`?

etienne-lms replied:
Yes, I agree. This is indeed confusing. Your proposal about renaming `cache_maintenance_l2()` into `cache_maintenance_outer_l2()` is nice. i will apply it.

jforissier replied:
OK, good. Please make the rename a separate patch, before this one (or after, as you prefer).

etienne-lms replied:
By the way, we currently do not define a `CFG_OUTER_CACHE` configuration directive, as only PL310 is supported as outer cache and `CFG_PL310` cover the case. I prefer to keep relying on `CFG_PL310,` rather than introducing an extra (and almost useless) `CFG_OUTER_CACHE.` Do you agree ?

jforissier replied:
Yes, let's just keep `CFG_PL310`.

jforissier commented:
Looks good to me.
`Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`

jenswi-linaro commented:
The outer cache could just as well be a level 3 cache (dependent on the number of inner caches), which level the outer cache represent isn't valid here.

While we're at it could I suggest aligning the a32 implementation with the a64 implementation, at least featurewise (deal with all levels of the inner cache, not assuming there's only one level)? I suppose the v7 implementation used in (or prepared for) ARM-TF would be nice to import.

jenswi-linaro commented:
Switching implementation for v7 cache maintenance could be done in another PR.

jforissier commented:
@jenswi-linaro so for the last commit in particular, are you recommending `s/cache_maintenance_outer_l2/cache_maintenance_outer/`? I'm OK with such a change, and as you said it could be a first step before `s/cache_maintenance_l1/cache_maintenance_inner/` (if I understand correctly).

etienne-lms commented:
I think current `cache_maintenance_l1()` acts on level1 cache, not level2 even if it is inner cache.

etienne-lms commented:
I corsscheck and I feel i was wrong. `cache_maintenance_l1()` acts on data cache, all levels of the unified cache.

jforissier commented:
@etienne-lms : @jenswi-linaro is *never* wrong ;)

jenswi-linaro commented:
@jforissier yes about the outer cache, with the flattery you're exaggerating. :-)
 
@etienne-lms good, then the a32 inner cache maintenance implementation is OK.

etienne-lms commented:
Je l'aurai, un jour. je l'aurai !  *  


(* french stupid commercial joke)

etienne-lms commented:
As i am renaming `cache_maintenance_l2()`, i can straight move to `cache_maintenance_inner/outer()`, as jerome suggested.

jenswi-linaro commented on core/arch/arm/mm/core_mmu.c:
```diff
@@ -648,7 +648,7 @@ int core_tlb_maintenance(int op, unsigned int a)
 }
 
-unsigned int cache_maintenance_l1(int op, void *va, size_t len)
+unsigned int cache_maintenance_l1(enum cache_op op, void *va, size_t len)
```
Please change the return type also.

etienne-lms replied:
right!

jenswi-linaro commented on core/arch/arm/mm/core_mmu.c:
```diff
@@ -690,7 +690,8 @@ unsigned int cache_maintenance_l1(int op, void *va, size_t len)
 
 #ifdef CFG_PL310
-unsigned int cache_maintenance_l2(int op, paddr_t pa, size_t len)
+unsigned int cache_maintenance_outer_l2(enum cache_op op,
```
Please change the return type also.

etienne-lms replied:
yes, thanks.

jforissier commented:
There is a typo in the commit message of your last update (`s/cache_op_outer/cache_op_inner/`).
Other than that, my R-b still applies.

etienne-lms commented:
Rebased (last commit is split in 3), tags applied (jens, I added your tag, I hope i did not wrong)
Travis was only complaining on a URL exceeding 80 chars (not fixed) and on space/tab (fixed).

Tested on b2260 (through the SDP test setup I must admit).

etienne-lms commented:
@jforissier : I fixed the typo in the commit message you pointed out.

etienne-lms commented:
(updated with jerome sole tag)

jenswi-linaro commented on core/include/tee/svc_cache.h:
```diff
@@ -31,7 +31,8 @@
 #ifdef CFG_CACHE_API
-TEE_Result syscall_cache_operation(void *va, size_t len, unsigned long op);
+TEE_Result syscall_cache_operation(void *va, size_t len,
+				   enum utee_cache_operation op);
```
I'd rather keep this as `unsigned long` instead of `enum utee_cache_operation` for two reasons, it's still `unsigned long` in `lib/libutee/include/utee_syscalls.h` and then there's also the comment at the beginning of that .h file.

etienne-lms replied:
I see your point. Ok to restore `unsigned long op`.

jenswi-linaro commented on core/arch/arm/include/mm/core_mmu.h:
```diff
@@ -375,23 +361,36 @@ enum teecore_tlb_op {
+	ICACHE_AREA_INVALIDATE,
+	DCACHE_CLEAN_INV,
+	DCACHE_AREA_CLEAN_INV,
+	L2CACHE_INVALIDATE,
```
Do we need these `L2CACHE_` enums? The outer cache is a data cache.

etienne-lms replied:
No, we don't. That's only legacy. Ok to clean it and move to `DCACHE_xxx` for both inner and outer cache data maintenance operations.

jenswi-linaro commented on core/arch/arm/tee/svc_cache.c:
```diff
@@ -60,5 +59,5 @@ TEE_Result syscall_cache_operation(void *va, size_t len,
 		return TEE_ERROR_ACCESS_DENIED;
 
-	return cache_operation(op, va, len);
+	return cache_operation((enum utee_cache_operation)op, va, len);
```
Is this cast needed?

etienne-lms replied:
strictly speaking, `cache_operation()` expects an ID from enumerated `utee_cache_operation.` Hence i would rather cast here so that the call fulfils the function prototype. I considered  `cache_operation()` as a core internal routine while `syscall_cache_operation()` is the syscall handler, hence conforming the `unsigned long` type used in the ABI.

jenswi-linaro replied:
So what is the purpose of this cast? Does it have any technical advantage or is it purely cosmetic?

etienne-lms replied:
It is cosmetic. Could it help compiler in some way ? Keeping or removing this cast is a matter of taste ? (i feel confortable with either cases)

jenswi-linaro replied:
Unnecessary casting could hide some warning and also clutters the code in my opinion. I doubt you'll find much of this pattern in the Linux kernel.

etienne-lms replied:
ok, then i discard this change. thanks.

jenswi-linaro commented:
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

jforissier commented:
@etienne-lms my R-b's still apply.

etienne-lms commented:
rebased, tags applied.

etienne-lms commented:
expecting travis to fail only on comment where a URL exceeds 80 chars.
